book/src/04_4_interrupts.md

```diff
@@ -1,5 +1,72 @@
 # Interrupts
 
-TODO link isr exercise
-- button press -> log message
+The goal of this exercise is to handle a button interrupt, so that the words "button pushed" get logged, if the `BOOT` button is pushed. 
+This exercise involves working with C bindings to the esp-idf-sys and other unsafe operations, as well as non-typical rust documentation. In a first step we will go line by line to build this interrupt handler. 
+
+You can find a skeleton code for this exercise in `advanced/button-interrupt/exercise/src/main.rs.`
+You can find the solution for this exercise in advanced/button-interrupt/solutuion/src/main.rs.
+
+TODO Add points where it's safe to build to rule out basic mistakes.
+TODO why are some functions called with some(...), esp!(...) or "normal"?
+
+## Tasks
+
+1. Configure the button (GPIO 9) with a c struct [`gpio_config_t`](https://esp-rs.github.io/esp-idf-sys/esp_idf_sys/struct.gpio_config_t.html)the following settings:
+    - input mode
+    - pull up
+    - interrupt on positive edge
+  
+  TODO add the variable names for the settings
+
+2. Write the configuration into the register with [`unsafe extern "C" fn gpio_config`](https://esp-rs.github.io/esp-idf-sys/esp_idf_sys/fn.gpio_config.html). This needs to happen in the unsafe block. To make these FFI calls we can use the macro `esp!($Cfunktion)`.
+
+
+3. Install the global interrupt handler with [`unsafe extern "C" fn gpio_install_isr_service`](https://esp-rs.github.io/esp-idf-sys/esp_idf_sys/fn.gpio_install_isr_service.html). This function needs `const ESP_INTR_FLAG_IRAM` as argument.
+
+
+4. Create a `static mut` that holds the state of the queue.
+
+```rust
+static mut EVENT_QUEUE: Option<QueueHandle_t> = None;
+```
+
+5. Create the event queue using [`pub unsafe extern "C" fn xQueueGenericCreate`](https://esp-rs.github.io/esp-idf-sys/esp_idf_sys/fn.xQueueGenericCreate.html)
+
+```rust
+EVENT_QUEUE = Some(xQueueGenericCreate(QUEUE_SIZE, ITEM_SIZE, QUEUE_TYPE_BASE));
+```
+
+6. Add a function to the RAM that determines what the interrupt handler does.
```

Comment:
```suggestion
6. Add a function which that will be called whenever there is a GPIO interrupt on our button pin. We put this function in a special block of RAM (`iram0`), so it will still be available even if the external flash is busy doing something else (like filesystem work). The function needs to get the queue handle from `EVENT_QUEUE` and call the `xQueueGiveFromISR` function with a `std::ptr::null_mut()` - the objects in our queue are of size zero, so we don't actually need a 'thing' to put on the queue. Instead, the act of pushing a 'nothing' is enough to wake up the other end!
```